flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkBatchRuleSets.scala

```diff
@@ -222,7 +222,8 @@ object FlinkBatchRuleSets {
   val JOIN_COND_EQUAL_TRANSFER_RULES: RuleSet = RuleSets.ofList((
     RuleSets.ofList(JoinConditionEqualityTransferRule.INSTANCE).asScala ++
       PREDICATE_SIMPLIFY_EXPRESSION_RULES.asScala ++
-      FILTER_RULES.asScala
+      FILTER_RULES.asScala ++
+      RuleSets.ofList(JoinDeriveNullFilterRule.INSTANCE).asScala
```

Comment:
It's better we do not put `JoinDeriveNullFilterRule`  into `JOIN_COND_EQUAL_TRANSFER_RULES` collection, instead I suggest we can add a program into  `JOIN_REWRITE` stage after `JOIN_COND_EQUAL_TRANSFER_RULES`